src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1561,14 +1576,17 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard>& shard)
 std::uint32_t
 DatabaseShardImp::shardBoundaryIndex() const
 {
+    auto const validIndex = app_.getLedgerMaster().getValidLedgerIndex();
+
+    if (validIndex < earliestLedgerSeq())
+        return validIndex;
+
     // Shards with an index earlier than the recent shard boundary index
     // are considered historical. The three shards at or later than
     // this index consist of the two most recently validated shards
     // and the shard still in the process of being built by live
     // transactions.
-    return NodeStore::seqToShardIndex(
-               app_.getLedgerMaster().getValidLedgerIndex(), ledgersPerShard_) -
-        1;
+    return seqToShardIndex(validIndex) - 1;
```

Comment:
I think you're mixing metaphors here. The `return validIndex` is returning a ledger index. The `return seqToShardIndex(validIndex) - 1` is returning a shard index. I'm guessing the idea was to avoid returning a negative / underflowed value, but that first return now doesn't make sense. Does it make more sense to `return 0` if the ledger index is too small?

Reply:
The idea was to avoid passing an index that's too early to `seqToShardIndex` which would cause an `assert` to fail. But I agree it makes more sense to just return zero. Done